pkg/runtime/serializer/codec_factory.go

```diff
@@ -352,10 +340,18 @@ func (c DirectCodec) EncodeToStream(obj runtime.Object, stream io.Writer, overri
 		return err
 	}
 	kind := obj.GetObjectKind()
+	oldGVK := kind.GroupVersionKind()
 	kind.SetGroupVersionKind(gvk)
+	defer kind.SetGroupVersionKind(oldGVK)
```

Comment:
It may be better not to defer here to avoid an allocation (this will be in the critical path for some code).  

```
err = c.Serializer.EncodeToStream(obj, stream, overrides...)
kind.SetGroupVersionKind(oldGVK)
return err
```

I eliminated the defer in another PR for versioning codecs.